### ISSUE TITLE
conflicting values when multiple traits patch an array

### ISSUE THREAD

**Issue by zou2699** [oam-dev/kubevela#1913]
**Describe the bug**

I have two traits, one provider redis secret, another provider MySQL secret, each of them have a patch template that patch  envFrom array to the component.

**To Reproduce**

1. The YAML files of Component/Trait I used.

   rds trait

   ```yaml
   # Code generated by KubeVela templates. DO NOT EDIT.
   apiVersion: core.oam.dev/v1beta1
   kind: TraitDefinition
   metadata:
     name: aliyun-rds
     namespace: vela-system
     annotations:
   spec:
     appliesToWorkloads:
       - webservice
       - worker
     schematic:
       cue:
         template: |
           patch: {
           	spec: template: spec: {
           		// +patchKey=name
           		containers: [{
           			name: context.name
           			// +patchKey=name
           			envFrom: [
           				{
           					secretRef: {
           						name: parameter.secret.name
           					}
           				},
           			]
           		}]
           	}
           }
           
           parameter: {
           	secret: {
           		name: string
           	}
           }
   ```

   redis trait

   ```yaml
   # Code generated by KubeVela templates. DO NOT EDIT.
   apiVersion: core.oam.dev/v1beta1
   kind: TraitDefinition
   metadata:
     name: aliyun-redis
     namespace: vela-system
   spec:
     appliesToWorkloads:
       - webservice
       - worker
     schematic:
       cue:
         template: |
           patch: {
           	spec: template: spec: {
           		// +patchKey=name
           		containers: [{
           			name: context.name
           			// +patchKey=name
           			envFrom: [
           				{
           					secretRef: {
           						name: parameter.secret.name
           					}
           				},
           			]
           		}]
           	}
           }
           
           parameter: {
           	secret: {
           		name: string
           	}
           }
   ```

   

2. The YAML file of Application I applied.

   ```yaml
   apiVersion: core.oam.dev/v1beta1
   kind: Application
   metadata:
     name: nginx
   spec:
     components:
       - name: nginx
         type: webservice
         properties:
           image: nginx:1.19
           port: 80
         traits:
           - type: aliyun-rds
             properties:
               secret:
                 name: nginx-rds
           - type: aliyun-redis
             properties:
               secret:
                 name: nginx-redis
   
   
   ```

   

3. Other operations I have done.

  ```
  Events:
    Type     Reason        Age              From         Message
    ----     ------        ----             ----         -------
    Normal   Parsed        5s (x3 over 8s)  Application  Parsed successfully
    Warning  FailedRender  5s (x3 over 8s)  Application  evaluate base template component=nginx app=nginx: cue: marshal error: spec.template.spec.containers.0.envFrom.0.secretRef.name: from source
  
  ```

  debug
  https://github.com/oam-dev/kubevela/blob/v1.0.7/pkg/appfile/appfile.go#L302
  the base  show it is conflicting values , but why not append the value to the envFrom array

  ```
  apiVersion: "apps/v1"
  kind:       "Deployment"
  spec: {
  	selector: {
  		matchLabels: {
  			"app.oam.dev/component": "nginx"
  		}
  	}
  	template: {
  		spec: {
  			// +patchKey=name
  			containers: [{
  				name:  "nginx"
  				image: "nginx:1.19"
  				ports: [{
  					containerPort: 80
  				}, ...]
  				// +patchKey=name
  				envFrom: [{
  					secretRef: {
  						name: _|_ // conflicting values "nginx-rds" and "nginx-redis"
  					}
  				}, ...]
  			}, ...]
  		}
  		metadata: {
  			labels: {
  				"app.oam.dev/component": "nginx"
  			}
  		}
  	}
  }
  
  ```

  

**Expected behavior**

it should append the value to the array, like this

```
apiVersion: "apps/v1"
kind:       "Deployment"
spec: {
	selector: {
		matchLabels: {
			"app.oam.dev/component": "nginx"
		}
	}
	template: {
		spec: {
			// +patchKey=name
			containers: [{
				name:  "nginx"
				image: "nginx:1.19"
				ports: [{
					containerPort: 80
				}, ...]
				// +patchKey=name
				envFrom: [{
					secretRef: {
						name:  "nginx-rds"
					},
					{
					secretRef: {
						name: "nginx-redis"
					}
				}, ...]
			}, ...]
		}
		metadata: {
			labels: {
				"app.oam.dev/component": "nginx"
			}
		}
	}
}
```

**Screenshots**

**KubeVela Version**

1.0.7

**Cluster information**

- Kubernetes Version  1.18.8

**Additional context**

**Comment by leejanee** [comment]
The element of the `envFrom`  array  is 
```yaml
{
        					secretRef: {
        						name: parameter.secret.name
        					}
        				}
```
,it does not contain the name field

**Comment by zou2699** [comment]
but how can I achieve this function
```yaml
    spec:
      containers:
      - envFrom:
        - secretRef:
            name: nginx-redis   # append
        - secretRef:
            name: nginx-rds    # append
        image: nginx:1.19
        imagePullPolicy: IfNotPresent
        name: nginx

```

**Comment by leejanee** [comment]
@zou2699 patchKey does not support value such as "a.b" now.

But,I think this is a general appeal. I will submit a pr to support it.

**Comment by wonderflow** [comment]
It was fixed in #1916 and you should write like this:

```
apiVersion: "apps/v1"
kind:       "Deployment"
spec: {
	selector: {
		matchLabels: {
			"app.oam.dev/component": "nginx"
		}
	}
	template: {
		spec: {
			// +patchKey=name
			containers: [{
				name:  "nginx"
				image: "nginx:1.19"
				ports: [{
					containerPort: 80
				}, ...]
				// +patchKey=secretRef.name
				envFrom: [{
					secretRef: {
						name:  "nginx-rds"
					},
					{
					secretRef: {
						name: "nginx-redis"
					}
				}, ...]
			}, ...]
		}
		metadata: {
			labels: {
				"app.oam.dev/component": "nginx"
			}
		}
	}
}
```
